amido/stacks-dotnet
# Feature/jwt bearer authn

[2531-Auth0 Implementation (.Net)]

#### 📲 What

Added JWT bearer token authentication to .NET Core Web API

#### 🤔 Why
		
To enable users of the Stacks .NET Core template to easily authenticate via OAuth 2.0.
		
#### 🛠 How
		
1. Configuration of JWT bearer authentication middleware
2. Configuration of Swagger API documentation to allow OAuth 2.0/bearer token authentication
3. Custom authorization provider to enable/disable [Authorize] attributes via configuration switch
4. Template stub HTTP message handler for component testing without passing traffic to OAuth 2.0 provider
5. Changes to component testing to test for 401/403 responses using test bearer tokens

#### 👀 Evidence
		
Screenshots / external resources / links / etc.
Link to documentation updated with changes impacted in the PR can be found here: https://amidodevelopment.atlassian.net/wiki/spaces/TEC/pages/1156448528/Application+Security+Placeholder
		 
#### 🕵️ How to test

1. Setup an authorization server and application in your chosen OAuth 2.0 provider (e.g. Auth0)
2. Update the `JwtBearerAuthentication` section in the `appsettings.json` file with the relevant OAuth 2.0 values and set `Enabled` to `true`
3. Deploy and run the application
4. Using the Swagger authentication create test users and generate tokens for them
5. Add the generated tokens (and corresponding user ids) into the `AuthTokenFixture.cs` file
6. In the browser navigate to the `/.well-known/openid-configuration` and `/.well-known/jwks.json` endpoints for your chosen OAuth 2.0 provider
7. Paste the responses of the two endpoints into the corresponding methods in file `StubJwtBearerAuthenticationHttpMessageHandler.cs`
8. Copy the `Audience`, `Authority`, `OpenApi.AuthorizationUrl`, `OpenApi.ClientId` and `OpenApi.TokenUrl` from appsettings.json to `CustomAutoDataAttribute.cs` and set `Enabled` to `true`
9. Run the component tests
10. Tests should all pass


#### ✅ Acceptance criteria Checklist
